File: jpms/wwwroot/js/boot-screen.js
// The boot screen: the jewel that fills the window from the first byte of HTML until the app has
// something real to show.
//
// It used to live INSIDE #app, which meant Blazor wiped it the instant it rendered its first
// frame — long before the session had been checked. The user watched a large centred jewel, then
// the chrome snapping in around a second, smaller jewel somewhere else on the page, then the
// page. Three states for one wait. Now the boot screen is a sibling overlay that stays up, and
// the app takes it down (BootScreen.DismissAsync, from ApprovedSessionGate, LandingLayout and
// ReportingErrorBoundary) at the moment it can replace it with content. The handover is a 200ms
// crossfade onto a screen that is already drawn.
//
// The failsafe matters more than the fade: a veil nobody removes is a dead app. Once Blazor has
// rendered anything at all into #app, a timer starts, and the boot screen goes whether or not the
// app remembered to dismiss it. It is deliberately longer than a cold start of the API: while the
// boot screen is up, the page gate's own jewel is drawn invisibly beneath it, and a failsafe that
// fires mid-session-check swaps one jewel for another — the very flicker this overlay exists to
// remove. Twelve seconds is past the slowest cold session check we see, and still short enough
// that a page which forgot to dismiss is usable rather than dead.
window.jpmsBoot = {
    FAILSAFE_MS: 12000,
    FADE_MS: 200,

    dismiss() {
        const boot = document.getElementById('boot');
        if (!boot || boot.dataset.going) return;
        boot.dataset.going = 'true';
        boot.style.transition = 'opacity ' + this.FADE_MS + 'ms ease';
        boot.style.opacity = '0';
        setTimeout(() => boot.remove(), this.FADE_MS);
    },

    // Boot itself failed — there is no app coming. The overlay stays: the ring stops, the jewel
    // holds still, and the #boot-failed template (index.html) replaces the "Loading" line.
    fail() {
        const boot = document.getElementById('boot');
        const failed = document.getElementById('boot-failed');
        if (!boot || !failed || boot.dataset.going) return;
        boot.querySelector('[data-boot-ring]')?.remove();
        boot.querySelector('[data-boot-jewel]')?.classList.remove('animate-jewel-pulse');
        boot.querySelector('[data-boot-message]')?.replaceWith(failed.content.cloneNode(true));
    }
};

(function () {
    const app = document.getElementById('app');
    if (!app) return;
    const armFailsafe = () => setTimeout(() => window.jpmsBoot.dismiss(), window.jpmsBoot.FAILSAFE_MS);
    if (app.childElementCount > 0) { armFailsafe(); return; }
    const watch = new MutationObserver(() => {
        if (app.childElementCount === 0) return;
        watch.disconnect();
        armFailsafe();
    });
    watch.observe(app, { childList: true });
})();

File: jpms/Services/BootScreen.cs
namespace Jewel.JPMS.Services;

/// <summary>
/// The app's side of the boot overlay in wwwroot/index.html (js/boot-screen.js): the jewel that
/// covers the window from the first byte of HTML until the app has something real to show.
/// </summary>
public static class BootScreen
{
    private static bool isDown;

    /// <summary>
    /// Takes the boot overlay down. Idempotent per document — the overlay exists once, so after
    /// the first successful call every later one (a page gate on each in-app navigation) is free.
    ///
    /// Best effort, never throws: the .NET assemblies and index.html are fetched separately, so a
    /// client can be running today's code against yesterday's shell — one that predates
    /// <c>jpmsBoot</c>, or a future one that renames it. A missing dismiss must not take the page
    /// render down with it (JPMS-4BF13E, 2026-09-08): the old shell kept its boot mark inside #app,
    /// which Blazor has already wiped, and the new shell's failsafe removes the overlay anyway.
    /// It is also called from the error boundary, where a second failure must not mask the first.
    /// </summary>
    public static async Task DismissAsync(IJSRuntime js)
    {
        if (isDown) return;
        isDown = true;
        try
        {
            await js.InvokeVoidAsync("jpmsBoot.dismiss");
        }
        catch (Exception)
        {
        }
    }
}
